docs/advanced/aml.rst

```diff
@@ -0,0 +1,203 @@
+    ...         inputs = np.array(idx, np.int64)
+    ...         nodes, edges, src_idx = graph_ops.sub_graph(
+    ...             g,
+    ...             inputs,
+    ...             edge_types=np.array(self.neighbor_edge_types, np.int64),
+    ...             num_hops=self.num_hops,
+    ...             self_loop=True,
+    ...             undirected=True,
+    ...             return_edges=True,
+    ...         )
+    ...         input_mask = np.zeros(nodes.size, np.bool)
+    ...         input_mask[src_idx] = True
+    ...
+    ...         feat = g.node_features(nodes, self.feature_meta, self.feature_type)
+    ...         label = g.node_features(nodes, self.label_meta, self.label_type).astype(np.int64)
+    ...         return {"nodes": np.expand_dims(nodes, 0), "feat": np.expand_dims(feat, 0), "labels": np.expand_dims(label, 0), "input_mask": np.expand_dims(input_mask, 0), "edges": np.expand_dims(edges, 0)}
+
+    >>> class GAT(BaseModel):
+    ...     def __init__(
+    ...         self,
+    ...         in_dim: int,
+    ...         head_num: List = [8, 1],
+    ...         hidden_dim: int = 8,
+    ...         num_classes: int = -1,
+    ...         ffd_drop: float = 0.0,
+    ...         attn_drop: float = 0.0,
+    ...     ):
+    ...         super().__init__(np.float32, 0, 0, None)
+    ...         self.num_classes = num_classes
+    ...         self.out_dim = num_classes
+    ...
+    ...         self.input_layer = GATConv(
+    ...             in_dim=in_dim,
+    ...             attn_heads=head_num[0],
+    ...             out_dim=hidden_dim,
+    ...             act=F.elu,
+    ...             in_drop=ffd_drop,
+    ...             coef_drop=attn_drop,
+    ...             attn_aggregate="concat",
+    ...         )
+    ...         layer0_output_dim = head_num[0] * hidden_dim
+    ...         assert len(head_num) == 2
+    ...         self.out_layer = GATConv(
+    ...             in_dim=layer0_output_dim,
+    ...             attn_heads=head_num[1],
+    ...             out_dim=self.out_dim,
+    ...             act=None,
+    ...             in_drop=ffd_drop,
+    ...             coef_drop=attn_drop,
+    ...             attn_aggregate="average",
+    ...         )
+    ...
+    ...     def forward(self, context: Dict[Any, np.ndarray]):
+    ...         nodes = torch.squeeze(context["nodes"])                # [N], N: num of nodes in subgraph
+    ...         feat = torch.squeeze(context["feat"])                  # [N, F]
+    ...         mask = torch.squeeze(context["input_mask"])            # [N]
+    ...         labels = torch.squeeze(context["labels"])              # [N]
+    ...         edges = torch.squeeze(context["edges"].reshape((-1, 2)))                # [X, 2], X: num of edges in subgraph
+    ...
+    ...         edges = np.transpose(edges)
+    ...
+    ...         # TODO This is not stable, when doing batch_size < graph size ends up with size < index values. use torch.unique to remap edges
```

Comment:
> This is not stable, when doing batch_size < graph size

Do you mean reverse? unstable when batch_size  is larger than graph size, because mapping is not unique anymore.